# apply Linear Regression

## load dataset

In [43]:
from sklearn.datasets import load_diabetes

import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [44]:
X,y = load_diabetes(return_X_y=True)

In [45]:
print(X.shape)
print(y.shape)

(442, 10)
(442,)


## train test split

In [46]:
X_train,X_test,y_train,y_test =train_test_split(X,y,test_size=0.2,random_state=2)

In [47]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(353, 10)
(353,)
(89, 10)
(89,)


## apply OLS
phala hum OLS apply kara raha ha taki fir baad ma hum kitna shai tha compare kar paya using gradient descant

In [48]:
reg_model = LinearRegression()
reg_model.fit(X,y)

LinearRegression()

In [49]:
# calcuate parameters values
print('intercept =',reg_model.intercept_)
print('coefficents =',reg_model.coef_)

intercept = 152.13348416289597
coefficents = [ -10.0098663  -239.81564367  519.84592005  324.3846455  -792.17563855
  476.73902101  101.04326794  177.06323767  751.27369956   67.62669218]


coefficents ma apka pass 10 values ha for each column(variable)

In [50]:
# calucate y predicate then r2 score
y_pred = reg_model.predict(X_test)
r2_score(y_test,y_pred)

0.45729681888868934

so 45 % of varience expalined by 10 varibles in the Y

# Our gradient class

## structure of it 

In [51]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        pass
    
    def predict(self,X_test):
        pass

In [52]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept = 0
        # lakin muja kesa pata hoga ki kitna coefficients ha 
        # tho X_train ka shape bataya ga number of coefficients
        self.coef_ = np.ones(X_train.shape[1])
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        pass
# lets check by creating our object
gd_object = GradientDescantRegressor()
gd_object.fit(X_train,y_train)


intercept : None
coefficents : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


## loop chalana ha 

### intercept 

slope of the intercept = $\frac{\partial L}{\partial B_0} = \frac{-2}{n} \sum_{i=1}^{n}(y_i - \hat{y_i})$           
Question : what is $y_i$ ?          
$y_i$ is basiclllay output column . meaning target variable  = hamara training data ma jo output column ha y_train ha           
Question : what is $\hat{y_i}$ ?            
$\hat{y_i}$ is apka model jo input data ka upar predication karega. example jab aap 353 rows deta ho tho apko
tho vo ek predication karega 353 predication kareaga        
deko tho             
$y_i$ is y_train = matrix of 353 rows and 10 columns        
$\hat{y_i}$ is y_test = matrix of 353 rows and 1 column         
Question : What is the meaning of $\frac{1}{n} \sum_{i=1}^{n} X{i}$ ya par $X_i = (y_i - \hat{y_i})$            
matlab aap iska mean nikal raha ho

observe the Sequence for slope calcuation 
- $(y_i - \hat{y_i})$ ek matrix ayaga with 353 rows
- ab hum us matrix ma mean nikalaga ga ek scaler ayega
- us scaler ko ab hum -2 sa multiply karega 
- fir us scaler ko learning rate ek scaler sa multiply kareag
- fir old intercept value sa us scaler ko subtract kareaga
- meaning our intercept parameter is always a scaler value

In [53]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        # lakin muja kesa pata hoga ki kitna coefficients ha 
        # tho X_train ka shape bataya ga number of coefficients
        self.coef_ = np.ones(X_train.shape[1])
        for i in range(self.epoches):
            # update all the coeff and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_
            slope_of_intercept  = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept) 
        
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        pass
# lets check by creating our object
gd_object = GradientDescantRegressor()
gd_object.fit(X_train,y_train)


intercept : 27.299789354156662
coefficents : [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


### coefficents

slope of the coefficent = $\frac{\partial L}{\partial B_m} = \frac{-2}{n} \sum_{i=1}^{n}(y_i - \hat{y_i})X_{mn}$

In [ ]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            # update all the coeff and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_
            slope_of_intercept  = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept) 

            slope_of_coefficents  =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
            self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents )
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        pass

gd_object = GradientDescantRegressor()
gd_object.fit(X_train,y_train)


intercept : 27.300032323395747
coefficents : [1.16382577 0.99950672 1.37912327 1.32841592 1.13278587 1.07264112
 0.78224268 1.22421099 1.42252845 1.23973457]


wow hamra gradient descant bhi bdya kam kar raha ha 

## predict 

In [ ]:
class GradientDescantRegressor:

    def __init__(self,learning_rate=0.001,epoches = 100):
        self.intercept_ = None
        self.coef_ = None
        self.learning_rate = learning_rate
        self.epoches = epoches
        
    def fit(self,X_train,y_train):
        # intialise the coeffs
        self.intercept_ = 0
        self.coef_ = np.ones(X_train.shape[1])
        self.coef_ = np.ones(X_train.shape[1])
        
        for i in range(self.epoches):
            # update all the coeff and the intercept
            y_hat = np.dot(X_train,self.coef_) + self.intercept_
            slope_of_intercept  = -2 * np.mean(y_train - y_hat)
            self.intercept_ = self.intercept_- (self.learning_rate * slope_of_intercept) 

            slope_of_coefficents  =  -2 * np.dot((y_train - y_hat),X_train)/X_train.shape[0]
            self.coef_ = self.coef_ - (self.learning_rate * slope_of_coefficents )
        print('intercept :',self.intercept_)
        print('coefficents :',self.coef_)
        
    def predict(self,X_test):
        return np.dot(X_test,self.coef_) + self.intercept_



In [60]:
gd_object = GradientDescantRegressor(epoches=1000,learning_rate=0.5)
gd_object.fit(X_train,y_train)

intercept : 152.01351687661833
coefficents : [  14.38990585 -173.7235727   491.54898524  323.91524824  -39.32648042
 -116.01061213 -194.04077415  103.38135565  451.63448787   97.57218278]


In [61]:
y_pred = gd_object.predict(X_test)

In [62]:
r2_score(y_test,y_pred)

0.4534503034722803

wow we verify in multiple linear regression that by using OLS or gradient descant you get the same result

today we learn humna scratch sa n-dim gradient descent lagana sika


normal gradient descant is batch gradient descant